File: frontend/src/utils/sleep.ts
/**
 * "Sen" — the night's vocabulary, and the one arithmetic this module performs.
 *
 * WHICH NIGHT AN ENTRY DESCRIBES. A sleep entry belongs to the night that
 * *ended* on the morning of its `date`: filled in on Friday morning, it
 * describes the night from Thursday to Friday, and its `date` is Friday. The
 * rule has to be written down somewhere because nothing about the data says it
 * — a row holding 23:40 and 06:50 is equally readable as Thursday's night or
 * Friday's, and the two answers put it in different weeks. It is stated once
 * here and once in `types/diet.ts`, next to the field itself.
 *
 * That choice follows from how the screen is used rather than from taste: the
 * form is filled in after waking, so the day the patient is looking at is the
 * day the entry lands on. It also keeps the module's one calendar rule intact —
 * `date` is always the local calendar day the patient is in, exactly as
 * `utils/days.ts` computes it and `core/days.py` mirrors it.
 *
 * WHAT IS COMPUTED HERE, AND NOTHING ELSE. `sleepDurationMinutes` is the only
 * derived value in the whole diet module. There is no sleep score, no average,
 * no comparison with other nights and no reading of how sleep relates to meals
 * or to mood — that last one is the analysis screen's job and a specialist's
 * conclusion, not a card's.
 */

import { addDays } from './days'

/**
 * How the night felt on waking.
 *
 * The mockup writes these as adjectives agreeing with a female patient
 * ("Wyspana", "Ociężale", "Spokojnie", "Z napięciem" — itself a mix of
 * adjectives and adverbs). Nouns instead: they name the same four states, read
 * identically to everybody, and put the row in one grammatical shape.
 */
export const WAKE_FEELING_OPTIONS = [
  { value: 'rested', label: 'Wyspanie' },
  { value: 'heavy', label: 'Ociężałość' },
  { value: 'calm', label: 'Spokój' },
  { value: 'tense', label: 'Napięcie' },
] as const

export type WakeFeeling = (typeof WAKE_FEELING_OPTIONS)[number]['value']

export const WAKE_FEELING_LABELS: Record<WakeFeeling, string> = Object.fromEntries(
  WAKE_FEELING_OPTIONS.map((option) => [option.value, option.label]),
) as Record<WakeFeeling, string>

/** The label to show, or null when the question went unanswered. */
export function wakeFeelingLabel(value: WakeFeeling | null | undefined): string | null {
  return value ? WAKE_FEELING_LABELS[value] : null
}

/**
 * The sleep-quality scale, 1 to 5.
 *
 * Numbers rather than named grades, which is what the mockup's detailed variant
 * draws. The alternative in the other variant ("Bardzo zła … Bardzo dobra")
 * asks the patient to judge their night in words; a number is the lighter
 * question, and this module's premise is that it describes rather than grades.
 */
export const SLEEP_QUALITY_VALUES = [1, 2, 3, 4, 5] as const

export type SleepQuality = (typeof SLEEP_QUALITY_VALUES)[number]

export function isSleepQuality(value: number): value is SleepQuality {
  return (SLEEP_QUALITY_VALUES as readonly number[]).includes(value)
}

const MINUTES_IN_DAY = 24 * 60

/**
 * What counts as a length of sleep: from one minute to 23 h 59 min.
 *
 * The ceiling is not a judgement about how long anybody may sleep — it is what
 * the two hours can express. A night is measured as the distance from one clock
 * reading forwards to the next, so every honest pair lands inside this range and
 * exactly one pair falls outside it: two identical hours, which measure a full
 * turn of the clock. That is a slip of the finger rather than a day-long sleep,
 * so it is answered with a request to correct it instead of a figure.
 */
export const MIN_SLEEP_MINUTES = 1
export const MAX_SLEEP_MINUTES = 23 * 60 + 59

/** 'HH:MM' → minutes since local midnight, or null for anything that is not one. */
function minutesSinceMidnight(time: string): number | null {
  const match = /^(\d{1,2}):(\d{2})$/.exec(time.trim())
  if (!match) return null
  const hours = Number(match[1])
  const minutes = Number(match[2])
  if (hours > 23 || minutes > 59) return null
  return hours * 60 + minutes
}

/**
 * How long the night lasted, in minutes — the module's only computed value.
 *
 * THE NIGHT CROSSES MIDNIGHT, which is the whole reason this is a function and
 * not a subtraction at the call site: 23:40 → 06:50 is 7 h 10 min, and the
 * subtraction on its own gives −16 h 50 min. Anything that does not run
 * strictly forwards within one day is read as having crossed midnight and gets
 * a day added.
 *
 * null when either hour is missing or unreadable — both fields are optional, so
 * "no length yet" is an ordinary state rather than a failure. The screen leaves
 * the figure out rather than printing a zero, for the same reason the diary
 * treats an untouched slider as unanswered instead of as nought.
 *
 * Also null when the result falls outside [MIN_SLEEP_MINUTES, MAX_SLEEP_MINUTES],
 * which in practice means two identical hours: 23:00 → 23:00 measures a full
 * turn of the clock, and a day-long sleep is a far less likely reading of it than
 * a mistyped digit. The screen answers that with a quiet request to correct one
 * of the hours (`sleepHoursCollide` is how it tells that case apart from a night
 * nobody has described yet) and still lets the entry be saved — no field in this
 * module blocks a save. The length simply is not computed.
 */
export function sleepDurationMinutes(
  fellAsleepAt: string | null,
  wokeUpAt: string | null,
): number | null {
  if (!fellAsleepAt || !wokeUpAt) return null
  const start = minutesSinceMidnight(fellAsleepAt)
  const end = minutesSinceMidnight(wokeUpAt)
  if (start === null || end === null) return null
  const forward = end - start
  const minutes = forward > 0 ? forward : forward + MINUTES_IN_DAY
  if (minutes < MIN_SLEEP_MINUTES || minutes > MAX_SLEEP_MINUTES) return null
  return minutes
}

/**
 * Whether the two hours are the same reading — the one pair that is answered
 * with a correction rather than with a figure.
 *
 * Kept apart from `sleepDurationMinutes` returning null because the screen has
 * to word two silences differently: a night nobody has filled in yet is waiting
 * for an answer, while this one has been answered in a way that cannot be
 * measured. Saying "policzy się, kiedy będą obie godziny" over two hours that
 * are both plainly there would read as the screen failing to notice them.
 */
export function sleepHoursCollide(
  fellAsleepAt: string | null,
  wokeUpAt: string | null,
): boolean {
  if (!fellAsleepAt || !wokeUpAt) return false
  const start = minutesSinceMidnight(fellAsleepAt)
  const end = minutesSinceMidnight(wokeUpAt)
  if (start === null || end === null) return false
  return start === end
}

/**
 * "7 h 10 min" — the mockup's own formatting.
 *
 * A whole number of hours drops the minutes ("7 h" rather than "7 h 0 min") and
 * a night under an hour drops the hours, because a unit showing nothing but a
 * zero reads as a field that failed to fill in.
 */
export function formatSleepDuration(minutes: number): string {
  const hours = Math.floor(minutes / 60)
  const rest = minutes % 60
  if (hours === 0) return `${rest} min`
  if (rest === 0) return `${hours} h`
  return `${hours} h ${rest} min`
}

/**
 * The first half of the label — the day the night began on, in the genitive and
 * *with its preposition*. Indexed the way `Date.getDay()` counts, Sunday first.
 *
 * Written out rather than taken from `toLocaleDateString`, which only ever
 * returns the nominative ("środa"): "noc z środa na czwartek" is not Polish.
 *
 * The preposition is part of the table rather than a literal at the call site
 * because Polish alternates it: "z" everywhere, but **"ze" before "środy"**,
 * where a bare "z" would collide with the ś. Wednesday is the only day of the
 * seven that needs it, which is exactly why a single hardcoded "z " would have
 * looked right in six tests out of seven.
 */
const WEEKDAY_NIGHT_FROM = [
  'z niedzieli',
  'z poniedziałku',
  'z wtorku',
  'ze środy',
  'z czwartku',
  'z piątku',
  'z soboty',
] as const

const WEEKDAY_ACCUSATIVE = [
  'niedzielę',
  'poniedziałek',
  'wtorek',
  'środę',
  'czwartek',
  'piątek',
  'sobotę',
] as const

/**
 * "noc z czwartku na piątek" — which night the entry on screen describes.
 *
 * Takes the morning the night ended on, i.e. the entry's own `date` as a Date,
 * and names the day before it and that day. See the note at the top of this
 * file for why the entry belongs to the later of the two.
 */
export function nightLabel(morning: Date): string {
  const evening = addDays(morning, -1)
  return `noc ${WEEKDAY_NIGHT_FROM[evening.getDay()]} na ${WEEKDAY_ACCUSATIVE[morning.getDay()]}`
}

File: backend/core/sleep.py
"""Sen — §09's second half, and the one rule about it nothing in the data states.

Out of the views like `core/activity.py` beside it, and like it this module sees
nothing but an `id_medical`.

**WHICH NIGHT A ROW DESCRIBES IS A CONVENTION, NOT A FACT IN THE DATA.** A night
carries the date of the *morning it ended on*: filled in on Friday, it is the
night from Thursday to Friday. A row holding 23:40 and 06:50 reads equally well
as either day, and the two answers put one night in two different weekly
reports — so the rule is written down here, at `models.DietSleep.entry_date`,
and at `frontend/src/types/diet.ts`'s own field. Changing it silently moves
nights between reports.

WHAT §09 DECIDES:

* **Five questions.** When you fell asleep, when you woke, how the night was on
  a 1-5 scale, how many times it was broken, and how you felt on waking. There
  is no sleep score, no weekly average, no seven-night chart and nothing
  relating sleep to meals or to mood — that last is a specialist's reading and
  belongs to the analysis screen, not to the form that collects it.
* **No field blocks a save** (§05), so every column but `awakenings` is
  nullable.
* **The length of the night is never stored.** It is the distance between the
  two hours, wrapping midnight, and `frontend/src/utils/sleep.ts` is the one
  place that arithmetic lives. A column would be a second answer free to
  disagree with the two hours beside it.

`WAKE_FEELINGS` are technical keys with the Polish labels in
`frontend/src/utils/sleep.ts` — the split `core/time_of_day.py` documents, and
the right way round for four words whose wording is a copy decision.
`core/tests/test_sleep.py` compares the two lists, names and order, in both
directions.
"""

from rest_framework import serializers

from .models import DietSleep

#: How the night is rated. Numbers rather than named grades, which is what the
#: mockup's other variant draws: "Bardzo zła … Bardzo dobra" asks the patient to
#: judge their night in words, and this module describes rather than grades.
SLEEP_QUALITY_VALUES = (1, 2, 3, 4, 5)

#: How somebody felt on waking, in the order §09 draws them. Technical keys; the
#: Polish labels are the frontend's.
WAKE_RESTED = 'rested'
WAKE_HEAVY = 'heavy'
WAKE_CALM = 'calm'
WAKE_TENSE = 'tense'

WAKE_FEELINGS = (WAKE_RESTED, WAKE_HEAVY, WAKE_CALM, WAKE_TENSE)

#: The most interruptions one night may record.
#:
#: Not a judgement about anybody's night — it is what a stepper somebody holds
#: down can otherwise write into a clinical record. Generous enough that no real
#: night meets it.
MAX_AWAKENINGS = 50


class SleepSerializer(serializers.Serializer):
    """What §09's "Zapisz" sends for one night.

    NOTHING IS REQUIRED, §05's rule again: a night that answers nothing is an
    ordinary row. The panel saves what is on the form, and the form starts
    empty.

    PUT REPLACES RATHER THAN MERGES, the same rule as `/api/diary/today/`, the
    supplement form and the meal form: the panel submits its whole state, so a
    field left out is an answer taken back rather than one left unchanged. A
    merge would make clearing an hour impossible from the only form that writes
    it.

    `wake_feeling` and `quality` are constrained rather than free, which is
    `0009`'s lesson: a plain `Serializer` drops an undeclared key *without an
    error*, and a value outside the vocabulary would be stored and then fail to
    render. Both are a 400 here.

    THE NIGHT IS NEVER AN INPUT. `entry_date` comes from the server's clock, so
    a date in the body reaches nothing — §09's rule is that a night whose
    morning has passed cannot be rewritten, and no URL names an older one.

    **`woke_up_at` earlier than `fell_asleep_at` is the ordinary case and is not
    refused**: the night crosses midnight. The one pair the frontend refuses is
    two identical hours (`sleepHoursCollide`), which measure a full turn of the
    clock and are a slip of the finger — and it refuses them by declining to
    print a length, not by blocking the save, because §05 says no field blocks
    one. So nothing here rejects a pair of hours either; the two ends agree.
    """

    fell_asleep_at = serializers.TimeField(required=False, allow_null=True)
    woke_up_at = serializers.TimeField(required=False, allow_null=True)
    quality = serializers.ChoiceField(
        choices=SLEEP_QUALITY_VALUES, required=False, allow_null=True,
    )
    awakenings = serializers.IntegerField(
        min_value=0, max_value=MAX_AWAKENINGS, required=False, default=0,
    )
    wake_feeling = serializers.ChoiceField(
        choices=WAKE_FEELINGS, required=False, allow_null=True, allow_blank=True,
    )

    def save_night(self, id_medical, day):
        """Write this night, replacing whatever was there.

        `update_or_create` against the `(id_medical, entry_date)` unique
        constraint, so a second save is an edit and not a second night — the
        same shape `activity.set_steps` gives a step count, and the same reason.
        """
        data = self.validated_data
        night, _ = DietSleep.objects.update_or_create(
            id_medical=id_medical, entry_date=day,
            defaults={
                'fell_asleep_at': data.get('fell_asleep_at'),
                'woke_up_at': data.get('woke_up_at'),
                'quality': data.get('quality'),
                'awakenings': data.get('awakenings') or 0,
                'wake_feeling': data.get('wake_feeling') or None,
            },
        )
        return night


def serialize_night(night, day):
    """One night, as `frontend/src/types/diet.ts`'s `DietSleepNight` reads it.

    `night` may be None, which is a morning nobody has answered for: the panel
    starts from exactly this shape, so an untouched night and a night saved with
    every field blank are one thing on the wire. They are also one thing in
    truth — `hasSleep` reads both as "not described" — which is why there is no
    attempt to tell them apart.

    The two hours travel as 'HH:MM', not as moments: nobody types seconds, and a
    precision nobody entered has no place in a clinical record.
    """
    if night is None:
        return {
            'date': day.isoformat(),
            'fell_asleep_at': None,
            'woke_up_at': None,
            'quality': None,
            'awakenings': 0,
            'wake_feeling': None,
        }

    return {
        'date': night.entry_date.isoformat(),
        'fell_asleep_at': (
            night.fell_asleep_at.strftime('%H:%M') if night.fell_asleep_at else None
        ),
        'woke_up_at': night.woke_up_at.strftime('%H:%M') if night.woke_up_at else None,
        'quality': night.quality,
        'awakenings': night.awakenings,
        'wake_feeling': night.wake_feeling or None,
    }


def load_night(id_medical, day):
    """One morning's night, or None when nobody has answered for it."""
    return DietSleep.objects.filter(id_medical=id_medical, entry_date=day).first()


def build_sleep_night(id_medical, day):
    """What §09's sleep panel draws — the night, or an empty shape for it."""
    return serialize_night(load_night(id_medical, day), day)


def load_history(id_medical, start, end):
    """Every night in a range, keyed by the morning it ended on.

    A morning nobody answered for is simply absent rather than an empty night:
    the report decides what an absent night means, and that is the one place the
    decision belongs. Bounded by the range, like `activity.load_history`.
    """
    rows = DietSleep.objects.filter(
        id_medical=id_medical, entry_date__gte=start, entry_date__lte=end,
    )
    return {
        row.entry_date.isoformat(): serialize_night(row, row.entry_date)
        for row in rows
    }


def first_entry_date(id_medical):
    """The earliest morning this patient answered for.

    An exact `MIN` rather than a scan of a capped history, for the reason
    `core/diet_reports.py` gives where it latches the week anchor.

    **A row whose every answer is blank does not count**, which is the one place
    this differs from the two diaries beside it: `build_sleep_night` answers
    with a full shape for a morning nobody has touched, so a panel that was
    merely opened and saved would otherwise anchor a patient's weeks on a night
    they never described. `has_night` is the same reading applied here.
    """
    rows = DietSleep.objects.filter(id_medical=id_medical).order_by('entry_date')
    for row in rows:
        if has_night(serialize_night(row, row.entry_date)):
            return row.entry_date
    return None


def has_night(night):
    """Whether a night was described at all.

    Mirrors `hasSleep` in `frontend/src/utils/dietReport.ts`, including its one
    subtlety: **`awakenings` counts as an answer only above zero.** Zero is both
    "an unbroken night" and the control's own starting value, and the two cannot
    be told apart, so the quieter reading wins — a night whose only non-default
    field is a zero it was born with is not a night somebody described.
    """
    return (
        night['fell_asleep_at'] is not None
        or night['woke_up_at'] is not None
        or night['quality'] is not None
        or night['wake_feeling'] is not None
        or night['awakenings'] > 0
    )
